biz.aQute.bndlib/src/aQute/bnd/osgi/Descriptors.java

```diff
@@ -654,4 +675,112 @@ public TypeRef getTypeRefFromPath(String path) {
 		return getTypeRef(path.substring(0, path.length() - 6));
 	}
 
+	public static String pathToFqn(String path) {
+		assert path.endsWith(".class");
+
+		StringBuilder sb = new StringBuilder();
+		int j = path.length() - 6;
+		for (int i = 0; i < j; i++) {
+			char c = path.charAt(i);
+			if (c == '/')
+				sb.append('.');
+			else
+				sb.append(c);
+		}
+		return sb.toString();
+	}
+
+	public static boolean isBinaryClass(String resource) {
+		return resource.endsWith(".class");
+	}
+
+	/**
+	 * Java really screwed up in using different names for the binary path and
+	 * the fqns. This calculates the simple name of a potentially nested class.
+	 *
+	 * @param resource ( segment '/')+ (name '$')* name '.class'
+	 * @return the last name
+	 */
+	public static String binaryToSimple(String resource) {
+		if (resource == null)
+			return null;
+
+		assert isBinaryClass(resource);
+
+		int end = resource.length() - 6;
+		int rover = end;
+		while (rover >= 0) {
+			char ch = resource.charAt(rover--);
+			if (ch == '$' || ch == '/')
+				break;
+		}
+		return resource.substring(rover + 1, end);
+	}
+
+	/**
+	 * Heuristic for a class name. We assume a segment with
+	 *
+	 * @param fqn can be a class name, nested class, or simple name
+	 * @return true if the last segment starts with an upper case
+	 */
+	public static boolean isClassName(String fqn) {
+		if (fqn.isEmpty())
+			return false;
+
+		int n = fqn.lastIndexOf('.') + 1;
+		if (n >= fqn.length())
+			return false;
+
+		char ch = fqn.charAt(n);
+
+		return Character.isUpperCase(ch);
+	}
+
+	/**
+	 * Return a 2 element array based on the fqn. The first element is the
+	 * package name, the second is the class name. Each can be absent, but not
+	 * both. The class name can be a nested class (will contain a '.' then)
+	 *
+	 * @param fqn a Java identifier name, either a simple class name, a
+	 *            qualified class name, or a package name
+	 * @return a Result with 2 element array with [package, class]
+	 */
+	public static Result<String[], String> determine(String fqn) {
```

Comment:
Would the return type be better as `Result<Map.Entry<String,String>,String>` or `Result<Pair<String,String>,String>`? Using `String[]` is pretty open and provides no clear evidence that the array contains exactly 2 strings.